### PR TITLE
Introduce Monolog support

### PR DESCRIPTION
**Pull request by Xenopathic**
Some new logging points too.

I tried to keep the chance of failure low - as soon as possible, a Logger is created pointing at the PHP error log. An error handler is also registered at this point, to catch any errors thrown by PHP and forward them to the logs. After the configuration is loaded, the error log handler is unregistered, and a stream handler pointing at `private/mesamatrix.log` is registered. From here on out any log messages go to that file instead of PHP's error log, but in case something during initialisation fails, it gets sent to the error log.

I thought about having configurable log handlers (email for example), but the added complexity wasn't worth it.

As for the console application - no logging to the file is done, but instead all logs just get forwarded to the console.

How to use:

```
\Mesamatrix::$logger->warning('...');
\Mesamatrix::$logger->notice('...');
\Mesamatrix::$logger->error('...');
...
```

Those are probably the most useful log functions, there are more of course.

### CODE REVIEW

File: lib/Git/ProcessBuilder.php [hunk 1]
```diff
@@ -27,7 +27,7 @@ public function __construct(array $arguments = array())
     {
         array_unshift($arguments, 'git');
         $gitDir = \Mesamatrix::path(\Mesamatrix::$config->getValue("info", "private_dir"))."/";
-        $gitDir .= \Mesamatrix::$config->getValue("git", "dir");
+        $gitDir .= \Mesamatrix::$config->getValue("git", "dir", "mesa.git");
```
Comment (MightyCreak) [inline lib/Git/ProcessBuilder.php]: Why did you add a default value to all the getValue()? Is it a restriction from the logger that I don't see or is it because you prefer that way?

File: lib/Parser/UrlCache.php [hunk 2]
```diff
@@ -38,13 +38,13 @@ public function __construct() {
      * This file is encoded in JSON.
      */
     public function load() {
-        $privateDir = \Mesamatrix::$config->getValue("info", "private_dir", "private");
+        $privateDir = \Mesamatrix::$config->getValue("info", "private_dir");
```
Comment (MightyCreak) [inline lib/Parser/UrlCache.php]: And why did you remove it here then?

File: lib/Config.php [hunk 3]
```diff
@@ -36,25 +35,6 @@ public function __construct($configDir) {
...
-            ini_set('display_errors', 1);
-            ini_set('error_reporting', E_ALL);
-            $this->debugMode = true;
-        }
-        else
-        {
-            ini_set('display_errors', 0);
-            ini_set('error_reporting', E_ALL & ~E_NOTICE & ~E_STRICT & ~E_DEPRECATED);
-            $this->debugMode = false;
-        }
```
Comment (MightyCreak) [inline lib/Config.php]: The problem I can see here is that we can't set the server's debug config anymore. I think we should keep the `ini_set()`s maybe based on the `log_level` in the config and set it in `Mesamatrix::init()` along with the other PHP ini config like the timezone.

Or maybe the new logger handles already that?

File: lib/base.php [hunk 4]
```diff
@@ -18,29 +18,45 @@
...
         self::$config = new \Mesamatrix\Config(self::$configDir);
 
-        date_default_timezone_set('UTC');
-    }
+        // register the log file
+        self::$logger->popHandler();
+        self::$logger->pushHandler(new StreamHandler(
+            self::path(self::$config->getValue('info', 'private_dir').'/mesamatrix.log'),
+            self::$config->getValue('info', 'log_level', Logger::WARNING)
+        ));
```
Comment (MightyCreak) [inline lib/base.php]: From what I understand, logs will be redirected only to a file. Maybe it would be interesting to have both logs, one stored in a file and one live in the terminal?

File: lib/Git/ProcessBuilder.php (hunk 1 above)
Comment (Xenopathic) [inline lib/Git/ProcessBuilder.php]: I just noticed that these values could do with a default. It's unrelated to the actual PR :laughing:

File: lib/Parser/UrlCache.php (hunk 2 above)
Comment (Xenopathic) [inline lib/Parser/UrlCache.php]: The way I see it, the private directory should always be set in the config, no exceptions. As such, there shouldn't be a default value.

File: lib/Config.php (hunk 3 above)
Comment (Xenopathic) [inline lib/Config.php]: I was in two minds about this, but I'll have a look at implementing those two INI prefs with the logger

File: lib/base.php (hunk 4 above)
Comment (Xenopathic) [inline lib/base.php]: "Live in the page"? You mean, display logs on the web page? That could be a really bad idea...

File: lib/Parser/UrlCache.php (hunk 2 above)
Comment (MightyCreak) [inline lib/Parser/UrlCache.php]: No worries, I was simply curious ;)  
It's fine by me.

File: lib/base.php (hunk 4 above)
Comment (MightyCreak) [inline lib/base.php]: A really bad idea for production, maybe not for debug.
Right now, when I code in PHP I heavily relying on what the browser (a.k.a. PHP) tells me. I'm trying to see if some PHP errors and warnings could be interesting after the PHP syntax parsing.

Comment (Xenopathic) [inline lib/base.php]: The problem is, if the page generation fails at some point due to an error, how do you print that error? The page is already broken... That's what log files are for :laughing:

File: lib/Config.php (hunk 3 above)
Comment (Xenopathic) [inline lib/Config.php]: Fixed

**Comment by MightyCreak** [comment]
:+1:
